### PR TITLE
Add LazyBatchNormXd

### PR DESCRIPTION
**Pull request by asi1024**
This PR implements UninitializedBuffer and LazyBatchnormXd based on #44538. (cc. @emcastillo and @albanD)

### CODE REVIEW

File: torch/nn/parameter.py [hunk 1]
```diff
@@ -138,3 +138,98 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             'explicitly manipulating `torch.nn.parameter.UninitializedParameter` '
             'objects. When using LazyModules Call `forward` with a dummy batch ' 
             'to initialize the parameters before calling torch functions'.format(func))
+
+
+class UninitializedBuffer(torch.Tensor):
+    r"""A parameter that is not initialized.
```
Comment (albanD) [inline torch/nn/parameter.py:144]: paremeter -> buffer

File: torch/nn/parameter.py [hunk 2]
```diff
@@ -138,3 +138,98 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
...
+
+    Unitialized Buffer is a a special case of :class:`torch.Tensor`
+    where the shape of the data is still unknown.
+
+    Unlikely a :class:`torch.Tensor`, uninitialized parameters
```
Comment (albanD) [inline torch/nn/parameter.py:149]: Unlike

File: torch/nn/parameter.py [hunk 3]
```diff
@@ -138,3 +138,98 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
...
+    hold no data and attempting to access some properties, like their shape,
+    will throw a runtime error. The only operations that can be performed on a uninitialized
+    parameter are changing its datatype, moving it to a different device and
+    converting it to a regular :class:`torch.Tensor`.
+    """
+    _allowed_methods = [
```
Comment (albanD) [inline torch/nn/parameter.py:155]: nit: can we share this list with the UninitializedParameter version above?

File: torch/nn/parameter.py [hunk 4]
```diff
@@ -138,3 +138,98 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             'explicitly manipulating `torch.nn.parameter.UninitializedParameter` '
             'objects. When using LazyModules Call `forward` with a dummy batch ' 
             'to initialize the parameters before calling torch functions'.format(func))
+
+
+class UninitializedBuffer(torch.Tensor):
```
Comment (albanD) [inline torch/nn/parameter.py:143]: This looks almost identical the parameter version above. i wonder if we could reduce the code duplication with a `UninitializedMixin` that is shared for both here?

File: test/test_nn.py [hunk 5]
```diff
@@ -14056,10 +14056,41 @@ def test_lazy_module_parameter(self):
...
+        new_module.register_buffer('test_buffer', UninitializedBuffer())
+        module.load_state_dict(new_module.state_dict())
+        module.load_state_dict(new_module.state_dict())
+        self.assertTrue(module.has_uninitialized_params())
+
     @suppress_warnings
     def test_lazy_module_jit(self):
         module = LazyModule()
         module.register_parameter('test_param', UninitializedParameter())
+        module.register_buffer('test_buffer', UninitializedBuffer())
```
Comment (albanD) [inline test/test_nn.py:14093]: I think this needs to be a different test to make sure that each of them are properly detected. Not the combination of them.

File: test/test_nn.py [hunk 6]
```diff
@@ -14068,6 +14099,7 @@ def test_lazy_module_jit(self):
     def test_lazy_share_memory(self):
         module = LazyModule()
         module.register_parameter('test_param', UninitializedParameter())
+        module.register_buffer('test_buffer', UninitializedBuffer())
```
Comment (albanD) [inline test/test_nn.py:14102]: Same as above, this needs to be two tests.

File: test/test_nn.py [hunk 7]
```diff
@@ -14245,6 +14277,115 @@ def test_lazy_conv_transpose3d_state(self):
...
+                input = torch.ones(*input_shape)
+                y = lazy_module(input)
+                self.assertIsInstance(lazy_module, cls)
+                self.assertNotIsInstance(lazy_module, lazy_cls)
+
+                num_features = input_shape[1]
+                module = cls(num_features, affine=affine, track_running_stats=track_running_stats)
+                expected = module(input)
+
+                if module.weight is not None:
```
Comment (albanD) [inline test/test_nn.py:14301]: Is there a reason why all these tests only check for shape and not for the computed value?
Do we want to check the computed value here? Or just the shape is enough?

File: torch/nn/modules/module.py [hunk 8]
```diff
@@ -1129,7 +1129,8 @@ def _load_from_state_dict(self, state_dict, prefix, local_metadata, strict,
                 # This is used to avoid copying uninitialized parameters into
                 # non-lazy modules, since they dont have the hook to do the checks
                 # in such case, it will error when accessing the .shape attribute.
-                is_param_lazy = isinstance(param, torch.nn.parameter.UninitializedParameter)
+                is_param_lazy = isinstance(param, (torch.nn.parameter.UninitializedParameter,
```
Comment (albanD) [inline torch/nn/modules/module.py:1132]: nit: the lazy package might provide a convenient function to do this to remove this "implementation detail" like `parameter.is_lazy(param)`.

**Review by albanD — COMMENTED** [review]
Thanks for the PR!
The batchnorm part looks good to me. Just small comments.

For the LazyBuffer, I think we can do some refactoring to simplify the code, see comment below.

**Comment by asi1024** [comment]
@albanD Thank you for the review! Could you take another look?

File: torch/nn/modules/module.py (hunk 8 above)
Comment (asi1024) [reply torch/nn/modules/module.py]: Now I implemented `UninitializedMixin` class simplified this line with `isinstance(..., UninitializedMixin)`.

File: test/test_nn.py (hunk 7 above)
Comment (asi1024) [reply test/test_nn.py:14315]: I fixed to check also computed values. Thanks!

File: torch/nn/parameter.py [hunk 9]
```diff
@@ -98,35 +84,35 @@ def materialize(self, shape, device=None, dtype=None):
...
     def share_memory_(self):
         raise RuntimeError(
-            'Can\'t share memory on an uninitialized parameter. '
+            'Can\'t share memory on an uninitialized parameter or buffer. '
             'Call `forward` to initialize the parameters before calling '
             '`module.share_memory()`.')
 
     def __repr__(self):
-        return 'Uninitialized parameter'
+        return '<' + self.__class__.__name__ + '>'
```
Comment (albanD) [inline torch/nn/parameter.py:104]: nit: now you can use f string to make these cleaner: `f'<{self.__class__.__name__}>'`!

File: torch/nn/parameter.py [hunk 10]
```diff
@@ -46,18 +46,7 @@ def __reduce_ex__(self, proto):
...
-    Unitialized Parameters are a a special case of :class:`torch.nn.Parameter`
-    where the shape of the data is still unknown.
-
-    Unlikely a :class:`torch.nn.Parameter`, uninitialized parameters
-    hold no data and attempting to access some properties, like their shape,
-    will throw a runtime error. The only operations that can be performed on a uninitialized
-    parameter are changing its datatype, moving it to a different device and
-    converting it to a regular :class:`torch.nn.Parameter`.
-    """
+class UninitializedMixin:
```
Comment (albanD) [inline torch/nn/parameter.py:49]: nit: `UninitializedMixin` -> `UninitializedTensorMixin` might be more precise on what it does?

File: torch/nn/modules/lazy.py [hunk 11]
```diff
@@ -226,8 +228,9 @@ def has_uninitialized_params(self: _LazyProtocol):
         # This is to avoid the JIT to track this parameter and force
         # custom modules __setstate__ to add it
         params = self._parameters.values()
-        for param in itertools.chain(params):
-            if isinstance(param, (UninitializedParameter)):
+        buffers = self._buffers.values()
+        for param in itertools.chain(params, buffers):
+            if isinstance(param, UninitializedMixin):
```
Comment (albanD) [inline torch/nn/modules/lazy.py:233]: There are quite a lot of calls to this isinstance check in a couple files and the "UninitializedMixin" might be hard to discover for people.
I still think that having a specific method to check this would make things easier to read and better for future modifications.

**Review by albanD — COMMENTED** [review]
Thanks for the update! It looks much better.

File: torch/nn/modules/lazy.py (hunk 11 above)
Comment (asi1024) [reply torch/nn/modules/lazy.py:233]: OK! I will add `is_lazy` function.

**Review by albanD — COMMENTED** [review]
The code looks good to me!
CI failures are real though, can you fix that? Let me know if you need help!

**Comment by asi1024** [comment]
@albanD I fixed mypy failure in 623ec45, but doc_test is still failing. Could you help me to pass the CI?

**Review by albanD — APPROVED** [review]
Hey!
Thanks for fixing the mypy issues.
The doc failure does look unrelated to me.

**Comment by albanD** [comment]
So the doc issues were real and broke master. Could you send a new PR with the same diff where we can check how to fix the doc?
